HotA map format support  - object mapping

IvanSavenko opened the pull request:
One more step towards hota h3m support.

List of changes in this PR:
- [x] Fix discovered issue with SoD h3m parsing
- [x] Allow defining map format parameters in gameSettings
- [x] Allow defining map format support per-format in gameSettings. Support of wog (h3m ID = 51) should not imply support for hota (h3m ID = 32)
- [x] Implement mapping system for town buildings. Integrate existing 'buildings5.json' into ID mapping system.
- [x] Implement mapping system for game entities, e.g. heroes, creatures, towns, etc
- [x] Implement mapping system for map objects 
- [x] Implement mapping system for map .def files
- [x] Implement hota map format mod as part of hota mod
- [x] Display hota maps in map selection screen with new icon
- [x] Add mapping for all hota objects

Known todo's for next PR(s):
- [x] (mod) HotA mod has validation issues
- [ ] (mod) Add loading of Cove special buildings
- [ ] Hero portrait mapping is broken
- [ ] Support loading of level 4 shrine from h3m
- [ ] Check loading of grave 
- [ ] Check loading of quest gates
- [ ] Add loading of new seer huts as new objects, with new strings
- [ ] Support for new victory/loss conditions from hota
- [ ] Ensure that all hota h3m maps can be correctly loaded & can be played
- [ ] Investigate support for hota campaigns support

Laserlicht commented on lib/mapping/CMap.h:
```diff
@@ -260,7 +260,7 @@ enum class EMapFormat: uint8_t
 	SOD   = 0x1c, // 28
```
Only as information, has nothing to do with the pull request: There is another official format.

For the Chronicles maps and campaigns. But I think this was mainly to prevent the Chronicles campaigns from being loaded into SoD. I don't think there is much difference.

https://github.com/potmdehex/homm3tools/blob/master/h3m/h3mlib/h3m_constants/h3m_format.h

https://github.com/potmdehex/homm3tools/blob/master/h3c/h3clib/h3clib.h

IvanSavenko replied:
Will add note in our source code, might be useful info in the future, once we can properly support datasets other than sod

IvanSavenko commented:
Extremely unstable & buggy for now, but at least some maps can now be loaded. 
Now stabilization...

![hota_h3m](https://github.com/vcmi/vcmi/assets/1576820/a28c3e5d-e7ba-40f6-8c91-8f9af3ac708d)

IvanSavenko commented:
Will probably merge this PR with ID mapping system as it for now, only new changes would be fixes for regression, if found.

Discovered too many issues that have to be dealt with one way or another, before we can properly support hota h3m in mod form.

Still planning hota h3m for 1.3, but not as part of this PR

Nordsoft91 commented:
Can be tested somehow now?

IvanSavenko commented:
Well, base game / non-hota maps should work as before.
If you wish to test hota maps then grab this PR and my PR in hota repository.

With these two PR's you should be able to:
- see all hota maps in scenario selection. However, maps with custom win/loss conditions don't have correct image & maps with predefined starting hero have invalid portrait.
- start default hota map, "Before the Storm", other maps might start too, but haven't checked all of them.
- visually, map seems to be correct, however there might be issues during actual gameplay. Especially seer huts, quest gates, potentially - grave and level 4 shrine.

Nordsoft91 commented:
Feedback, hopefully it's helpful:
- underground button is available on maps without underground (maybe unrelated to this pr, but regression)
- Suggest to split hota map support submod from hota maps (mod which contains actual maps). As user, I want to enable map support without having duplicated set of maps (if I use hota resources or if I have same maps in different language)
- rmg: some objects are placed unexpectedly:
<img width="223" alt="Снимок экрана 2023-05-28 в 03 08 26" src="https://github.com/vcmi/vcmi/assets/9308612/94f3591a-01b5-4c1c-8165-e8cfff5f885a">
<img width="176" alt="Снимок экрана 2023-05-28 в 03 10 31" src="https://github.com/vcmi/vcmi/assets/9308612/faa419ac-a562-4c87-af51-5e89a173cf4a">

- Only 22 out of 45 maps loaded successfully. Other have server crash
- Loaded successfully: A Cold Day in Hell, Before the storm, Blazing Archipelago, Face to Face, Fortune Favors Fools, Gambit, Glorious days, Gold Rush, Island of dust, King of the hill, Maelstorm, Marsh, Mysterious ways, Perfect equality, Pirate's utopia, Province, Strait of Storms (both), Talent Shortage, The Challenge, The Great Wasteland, With Sword and Fire
- Had a client crash while collecting obelisks on **Pirate's utopia** (piece from log, maybe related to absence on music, I don't have it)
```
[2023-May-28 03:37:24.539003] TRACE network - 	Applied on gs: 11TryMoveHero
[2023-May-28 03:37:24.539018] TRACE global - Del-ing music file Music/Water.mp3
[2023-May-28 03:37:24.539030] TRACE global - Loading music file Music/Water.mp3
[2023-May-28 03:37:24.539073] ERROR global - Failed to load music. setName=terrain_core:water	musicURI=
[2023-May-28 03:37:24.539098] ERROR global - Exception: Resource with name MUSIC/WATER and type SOUND wasn't found.
[2023-May-28 03:37:24.539118] WARN global - Warning: Cannot open Music/Water.mp3: music parameter was NULL
[2023-May-28 03:37:24.539135] TRACE global - Playing music file Music/Water.mp3
[2023-May-28 03:37:24.539145] ERROR global - Unable to play music (music parameter was NULL)
[2023-May-28 03:37:24.543587] ERROR global - vector
```
- Had a client crash while playing **The Great Wasteland** - crash at mouse hovering object on screenshot
<img width="306" alt="Снимок экрана 2023-05-28 в 03 53 32" src="https://github.com/vcmi/vcmi/assets/9308612/f0738e14-ca73-4d9f-b6b4-1f447c00793b">

- **Trial Of Flame** - empty (victory?) message
- **With Sword and Fire** - crash at event trigger, event is located after the bridge near to start
- Map Editor: I was able to open only Gambit, even Before the Storm isn't loaded (but I didn't try all maps)
- Map Editor: heroes positions are incorrect (Gambit)
<img width="196" alt="Снимок экрана 2023-05-28 в 04 11 01" src="https://github.com/vcmi/vcmi/assets/9308612/42f84998-62ab-4e22-932d-866afa1e9839">

IvanSavenko commented:
- split hota maps & map support submods - makes sense, although hota already has too many submods. Might be a good idea to merge some of them.
- new objects in RMG - found how they were "disabled" before, reverted change

Rest of issues - kind of known, or at least - expected. Will make some preparations & try to tackle hota h3m again.
